### PR TITLE
Batched dispatcher fixes and enhancements

### PR DESCRIPTION
**Pull request by zikaari**
Fixes event delivery issue when `analytics.track` is called right before
an action that causes page to unload, for example clicking a link or closing
a tab. The problem was caused by `beforeunload` handler being called before
`dispatch` which actually pushes the event to the buffer, since latter is part
of an async pipeline, unlike `beforeunload` sequence which is synchronous. The
issue is fixed by adding a `pageUnloaded` flag which gets set to true in the
`beforeunload` handler, so if `dispatch` gets called any time after that, it'd
know to send the event right away since there's no time to waste as page is about
to be unloaded.

Enhancements:

- Removed the `flushing` flag. It's value is changed from `true` to `false` synchronously
inside `flush`, so effectively the value of `flushing` is always `false` therefore serving
no actual purpose.

- Simplified the `buffer` type. The first member of the inner array serves no purpose, so
it's been flattened to simple 1D array.

- De-duplicated `fetch` calls into a common `sendBatch` function.

- Flush scheduler is now self-contained. Previoulsy the timer ID was managed by callers,
which means there was a potential for a caller to not check if flush has already been
scheduled. This could've led to "phantom" timers if `scheduleFlush` had been called multiple
times without checking.

- Window unload tests are now more realistic as they *trigger* the `beforeunload` event instead
calling the handler directly.

### CODE REVIEW

**Comment by zikaari** [comment]
Fixes #361 

We'll merge this once we get a greenlight from @mlstubblefield

File: src/plugins/segmentio/__tests__/batched-dispatcher.test.ts
Comment (chrisradek) [inline src/plugins/segmentio/__tests__/batched-dispatcher.test.ts:220]: Do we want errors in `dispatch` to actually fail the test? It looks like it would have before.

**Review by chrisradek — APPROVED** [review]
Overall I think this looks great! Love the refactoring you did and how you reduced the fetch duplication!

Just had a couple questions that shouldn't be considered blockers (unless we're actually seeing errors logged in the test now)

File: src/plugins/segmentio/__tests__/batched-dispatcher.test.ts
Comment (zikaari) [reply src/plugins/segmentio/__tests__/batched-dispatcher.test.ts:220]: This is to get rid of the eslint error "No unhandled promises".

The reason `await` has been removed is to deliberately test how well the batched dispatcher handles a lot of "pressure" at once (`await` would have been simulating events coming in one by one).

The test would still will fail because if `dispatch` fails to do its job, latter assertions won't pass either.

File: src/plugins/segmentio/batched-dispatcher.ts
Comment (zikaari) [reply src/plugins/segmentio/batched-dispatcher.ts:64]: It's a remnant of earlier experimentation I was doing. Good catch. Fixed up.

File: src/plugins/segmentio/batched-dispatcher.ts
Comment (zikaari) [reply src/plugins/segmentio/batched-dispatcher.ts:65]: Yes. Pooya confirmed it to me that our payloads can exceed that max limit.

**Comment by zikaari** [comment]
Tested by the customer and ready to be merged.

cc @chrisradek @pooyaj 

https://github.com/segmentio/analytics-next/issues/361#issuecomment-1053855542

**Review by pooyaj — APPROVED** [review]

**Pull request merged**
